File: dsa/medium/9. Merge Overlapping INtervals/merge_overlapping_intervals.cpp
#include <algorithm>
#include <vector>

std::vector<std::vector<int>>
mergeOverlappingIntervals(std::vector<std::vector<int>> intervals) {
  std::vector<std::vector<int>> sortedIntervals = intervals;
  std::sort(sortedIntervals.begin(), sortedIntervals.end(),
            [](std::vector<int> a, std::vector<int> b) { return a[0] < b[0]; });

  std::vector<std::vector<int> *> mergedIntervals;
  std::vector<int> *currentInterval = &sortedIntervals[0];
  mergedIntervals.push_back(currentInterval);

  for (auto &nextInterval : sortedIntervals) {
    int currentIntervalEnd = currentInterval->at(1);
    int nextIntervalStart = nextInterval.at(0);
    int nextIntervalEnd = nextInterval.at(1);

    if (currentIntervalEnd >= nextIntervalStart) {
      currentInterval->at(1) = std::max(currentIntervalEnd, nextIntervalEnd);
    } else {
      currentInterval = &nextInterval;
      mergedIntervals.push_back(currentInterval);
    }
  }

  std::vector<std::vector<int>> mergedIntervalsVector;
  for (auto &interval : mergedIntervals) {
    mergedIntervalsVector.push_back(*interval);
  }
  return mergedIntervalsVector;
}

File: dsa/medium/9. Merge Overlapping INtervals/merge_overlapping_intervals.go
package main

// import the sort, fmt and the math package
import (
	"math"
	"sort"
)

func MergeOverlappingIntervals(intervals [][]int) [][]int {
	if len(intervals) == 0 {
		return intervals
	}
	sortedIntervals := make([][]int, len(intervals))
	copy(sortedIntervals, intervals)
	sort.Slice(sortedIntervals, func(i, j int) bool {
		return sortedIntervals[i][0] < sortedIntervals[j][0]
	})

	var mergedIntervals [][]int
	mergedIntervals = append(mergedIntervals, sortedIntervals[0])
	currentInterval := sortedIntervals[0]

	for i := 1; i < len(sortedIntervals); i++ {
		currentIntervalEnd := currentInterval[1]
		nextIntervalStart := sortedIntervals[i][0]
		nextIntervalEnd := sortedIntervals[i][1]

		if currentIntervalEnd >= nextIntervalStart {
			currentInterval[1] = int(math.Max(float64(currentIntervalEnd), float64(nextIntervalEnd)))
		} else {
			mergedIntervals = append(mergedIntervals, sortedIntervals[i])
			currentInterval = sortedIntervals[i]
		}
	}
	return mergedIntervals
}
